operator/endpointgc/gc_test.go

```diff
@@ -0,0 +1,246 @@
+	cepWithOwnerNotRunningPod := createCiliumEndpoint("cep3", "ns")
+	cepWithOwnerNotRunningPod.OwnerReferences = []meta_v1.OwnerReference{createOwnerReference("Pod", "pod3")}
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithOwnerNotRunningPod, meta_v1.CreateOptions{})
+	// - CEP with no owner reference but with pod that is running
+	cepWithLegacyRunningPod := createCiliumEndpoint("cep4", "ns")
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithLegacyRunningPod, meta_v1.CreateOptions{})
+	// - CEP with no owner reference but with pod that is not running
+	cepWithLegacyNotRunningPod := createCiliumEndpoint("cep5", "ns")
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithLegacyNotRunningPod, meta_v1.CreateOptions{})
+	// - CEP with owner reference Cilium Node
+	cepWithOwnerCiliumNode := createCiliumEndpoint("cep6", "ns")
+	cepWithOwnerCiliumNode.OwnerReferences = []meta_v1.OwnerReference{createOwnerReference("CiliumNode", "node6")}
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithOwnerCiliumNode, meta_v1.CreateOptions{})
+	// - CEP with owner reference pod that doesn't exist
+	cepWithOwnerPodDoesntExist := createCiliumEndpoint("cep7", "ns")
+	cepWithOwnerPodDoesntExist.OwnerReferences = []meta_v1.OwnerReference{createOwnerReference("Pod", "pod7")}
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithOwnerPodDoesntExist, meta_v1.CreateOptions{})
+	// - CEP with owner reference cilium node that doesn't exist
+	cepWithOwnerCiliumNodeDoesntExist := createCiliumEndpoint("cep8", "ns")
+	cepWithOwnerCiliumNodeDoesntExist.OwnerReferences = []meta_v1.OwnerReference{createOwnerReference("CiliumNode", "node8")}
+	fakeClient.CiliumV2().CiliumEndpoints("ns").Create(context.Background(), cepWithOwnerCiliumNodeDoesntExist, meta_v1.CreateOptions{})
+
+	// Create Pods
+	// - pod that is running for cep2
+	fakeClient.Slim().CoreV1().Pods("ns").Create(context.Background(), createPod("pod2", "ns", true), meta_v1.CreateOptions{})
+	// - pod that is not running for cep3
+	fakeClient.Slim().CoreV1().Pods("ns").Create(context.Background(), createPod("pod3", "ns", false), meta_v1.CreateOptions{})
+	// - pod that is running for cep4
+	fakeClient.Slim().CoreV1().Pods("ns").Create(context.Background(), createPod("cep4", "ns", true), meta_v1.CreateOptions{})
+	// - pod that is not running for cep5
+	fakeClient.Slim().CoreV1().Pods("ns").Create(context.Background(), createPod("cep5", "ns", false), meta_v1.CreateOptions{})
+
+	// Create CiliumNodes
+	// - cilium node for cep6
+	fakeClient.CiliumV2().CiliumNodes().Create(context.Background(), createCiliumNode("node6"), meta_v1.CreateOptions{})
+}
+
+func waitForCEPs(t *testing.T, cepStore resource.Store[*cilium_v2.CiliumEndpoint], number int) {
+	if err := testutils.WaitUntil(func() bool {
+		return len(cepStore.List()) == number
+	}, waitTime*5); err != nil {
```

Comment:
I suggest to increase this a bit, maybe to a total of 5 or 10 seconds. This is to reduce flakyness in CI, where the performance characteristic is so unpredictable. Also, this won't introduce any fixed delay, since the condition is still checked every 10ms.

Reply:
Changed. I have a problem with a test for disabled endpoints gc.
```go
// wait for all CEPs created by prepareCiliumEndpoints
waitForCEPs(t, cepStore, 8)
// wait for potential GC
time.Sleep(500 * time.Millisecond)
// gc is disabled so no CEPs should be deleted
waitForCEPs(t, cepStore, 8)
```
It's hard to tell how long should we wait to make sure gc is really disabled.

But maybe it's fine to just remove this test.

Reply:
> It's hard to tell how long should we wait to make sure gc is really disabled.

That's right.
In this case, I don't think we can do much to improve this. Maybe what we can do is explicitly setting the gc interval with a value much smaller than the observation interval:

```diff
diff --git a/operator/endpointgc/gc_test.go b/operator/endpointgc/gc_test.go
index d0e23d52e5..d3acc2a8ee 100644
--- a/operator/endpointgc/gc_test.go
+++ b/operator/endpointgc/gc_test.go
@@ -112,6 +112,7 @@ func TestRegisterControllerWithCRDDisabled(t *testing.T) {
                cell.Metric(NewMetrics),
                cell.Provide(func() SharedConfig {
                        return SharedConfig{
+                               Interval:                 100 * time.Millisecond,
                                DisableCiliumEndpointCRD: true,
                        }
                }),
```

Unfortunately, this is still time-dependent, but (hopefully) should be fine in most cases and makes the test intent a bit clearer.